FSDP communication hook interface for NO_SHARD strategy

aovladi opened the pull request:
Fixes #79114

An implementation of a FSDP communication hook interface for a NO_SHARD strategy:
- `FullyShardedDataParallel.register_comm_hook(self, state: object, hook: callable)` checks current sharding strategy. If it is other that NO_SHARD, raises a runtime error. Otherwise, sets and shares a specified hook and its state with all submodules
- When FSDP is ready to communicate a gradient, checks if there is a registered hook, and calls it instead of all_reduce. Additionally, gradient pre and post devision are not performed if a hook is registered.

To test the interface, I've implemented a communication hook, that calls for `all_reduce`.

A  unittest:
- checks that is a sharding strategy is anything but NO_SHARD, a runtime error is raised
- checks that for a NO_SHARD case, model with registered all_reduce hook and without a hook work the same.
- checks for 2 types of FSDP models: with the wrapped first layer and without. (to make sure submodules have a hook registered)

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2822,7 +2824,7 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                     param.grad.data = param.grad.data.to(self.mixed_precision.reduce_dtype)
 
-                if self.gradient_predivide_factor > 1:
+                if self.gradient_predivide_factor > 1 and self.sharding_strategy != ShardingStrategy.NO_SHARD:
```
might be clearer if we just check that a hook is not registered

zhaojuanmao replied:
same, we should not change default logic here

aovladi replied:
Noted, adjusted the logic.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2899,10 +2901,20 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
-                            param.grad.div_(self.gradient_postdivide_factor)
+                        # if a communication hook was registered,
+                        # then a model has attribute ``communication_hook``
+                        if hasattr(self, "communication_hook"):
```
setting `self.comm_hook = None` in ctor might be cleaner, as we can just check `if self.comm_hook is not None` instead of hasattr check.

aovladi replied:
Got it. Adjusted logic in the new commit.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2899,10 +2901,20 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+                        else:
+                            # we skipped predivide earlier for NO_SHARD
+                            # this is a no-hook case, so need to predivide grads
+                            if self.gradient_predivide_factor > 1:
```
might be a bit confusing we're doing predivide in a different location depending on no_shard or not, we should possibly case the divide decision based on whether there is a hook registered or not

aovladi replied:
Noted. Refactored code accordingly.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3291,6 +3303,54 @@ def _check_rebuild_full_params(self, param: FlatParameter):
+                             that needs to be communicated across ranks. It then performs
+                             all neccessary processing and returns nothing.
+
+        """
```
awesome documentation!!

aovladi replied:
Thanks!

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3291,6 +3303,54 @@ def _check_rebuild_full_params(self, param: FlatParameter):
+                            and shared by all the gradient tensors on the worker.
+
+            hook (callable): Callable with the following signature:
+                             ``hook: Callable[torch.Tensor] -> None``:
```
mention what this parameter corresponds to specifically (gradient of what)

aovladi replied:
Clarified

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3291,6 +3303,54 @@ def _check_rebuild_full_params(self, param: FlatParameter):
+                    logger.warn(
+                        f"NOTE: {hook.__qualname__} will be shared across all submodules."
+                    )
+                submodule.communication_hook_state = state
```
as a sanity check, might be good to make sure we're not double-registering hooks on submodules

aovladi replied:
Added into the new version

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3291,6 +3303,54 @@ def _check_rebuild_full_params(self, param: FlatParameter):
+            # register same hook for root and all submodules
+            for submodule in self.fsdp_modules(self):
+                if submodule.check_is_root():
+                    logger.warn(
```
I would advocate for removing this warning and just inserting it into the documentation, it is expected behavior so not really worth raising a warning imo

aovladi replied:
Removed warning from a logger

rohan-varma commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,172 @@
+                        self.fc1 = wrap(nn.Linear(8, 16))
+                else:
+                    self.fc1 = nn.Linear(8, 16)
+                self.relu = F.relu
```
don't need this and just do F.relu in forward?

aovladi replied:
Refactored

rohan-varma commented on torch/distributed/algorithms/fsdp_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,49 @@
+        world_size
+    ):
+        self.process_group = process_group
+        self.world_size = world_size
```
self.world_size = dist.get_world_size(self.process_group)?

aovladi replied:
Refactored accordingly

rohan-varma commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,172 @@
+                torch.cuda.manual_seed(0)
+                super().__init__()
+                if has_wrapping:
+                    with enable_wrap(wrapper_cls=FSDP, device_id=torch.cuda.current_device()):
```
let's also test one layer deeper of wrapping just to be sure:

FSDP(
... params
FSDP(
...params
FSDP(
...params
)))

aovladi replied:
Got it. Added extra layer in an updated version.

rohan-varma reviewed (commented):
awesome work! Quick initial comments, will come back for thorough review

mrshenli commented on torch/distributed/algorithms/fsdp_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,49 @@
+import torch
```
Does it make sense if we start as a private/prototype feature (i.e., `_fsdp_comm_hooks`) first? Or is this planned to be a beta feature in the v1.13 release?

aovladi replied:
That is not planned, thanks for pointing this out. Moved to a private folder `_comm_hooks`

mrshenli commented on torch/distributed/algorithms/fsdp_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,49 @@
+
+class AllReduceState(object):
+    r"""
+    Stores parameters, needed to perform ``all_reduce`` algorithm
```
nit: not all args are documented.

aovladi replied:
Expanded docs with args' description.

mrshenli commented on torch/distributed/algorithms/fsdp_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,49 @@
+        self.gradient_postdivide_factor = self.world_size / self.gradient_predivide_factor
+
+    # Same as in `FullyShardedDataParallel` class.
+    # Required to perform gradient pre-and post division
```
"pre-and post division" -> "pre- and post-division"?

mrshenli commented on torch/distributed/algorithms/fsdp_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,49 @@
+def allreduce_hook(state: AllReduceState, grad: torch.Tensor):
+    r"""
+    This FSDP communication hook implements ``all_reduce`` algorithm
+    and neccessary pre-and post devision of gradients.
```
ditto

mrshenli commented on torch/distributed/algorithms/fsdp_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,49 @@
+    # Required to perform gradient pre-and post division
+    def _get_gradient_predivide_factor(self, world_size: int) -> float:
+        factor: int = 1
+        while world_size % factor == 0 and world_size / factor > factor:
+            factor *= 2
```
Let's copy the FSDP comment here:

```
    # setting two factors 'self.gradient_predivide_factor'
    # and 'self.gradient_postdivide_factor' to avoid underflow and overflow
```

@rohan-varma do you recall why this is sufficient to avoid overflow?

aovladi replied:
Added comments. Discussed this with Rohan offline

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3291,6 +3303,54 @@ def _check_rebuild_full_params(self, param: FlatParameter):
+        flexible hook to users where they can specify how FSDP aggregates gradients
+        across multiple workers.
+
+        This hook can be used to implement several algorithms like GossipGrad
```
citations?

aovladi replied:
added citation

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3291,6 +3303,54 @@ def _check_rebuild_full_params(self, param: FlatParameter):
+        parameter syncs while running Fully Sharded Data Parallel training.
+
+        .. warning::
+            FSDP only support communication hooks for a NO_SHARD strategy at this time.
```
nit: double quote `NO_SHARD` to show it as code.

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3291,6 +3303,54 @@ def _check_rebuild_full_params(self, param: FlatParameter):
             eod.param_order.append(param_index)
 
+    def register_comm_hook(self, state: object, hook: callable):
+        r"""
```
could you please build the docs and check if this is rendered properly?

aovladi replied:
I've checked docs build without issues, will verify how it is rendered in PR's docs

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3291,6 +3303,54 @@ def _check_rebuild_full_params(self, param: FlatParameter):
+        """
+        if self.sharding_strategy != ShardingStrategy.NO_SHARD:
+
+            raise RuntimeError(
```
this should be `NotImplementedError`?

aovladi replied:
Changed to `NotImplementedError`

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2899,10 +2901,20 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+                        # then a model has attribute ``communication_hook``
+                        if hasattr(self, "communication_hook"):
+                            self.communication_hook(self.communication_hook_state, param.grad)
+                        else:
```
This branch can be put into a default hook, so that two branch can merge into one?

aovladi replied:
Thank you for suggestion! Refactored code accordingly and added some tests

mrshenli reviewed (commented):
Hey @aovladi, thanks a lot for contributing this PR! I left some minor comments.

zhaojuanmao commented on torch/distributed/algorithms/fsdp_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,49 @@
+import torch
+import torch.distributed as dist
```
could we change the fsdp_comm_hooks to a general name like '_comm_hooks'? so that the DDP common hook can be merged here?

aovladi replied:
Thanks for the suggestion, moved to a new folder.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3291,6 +3303,54 @@ def _check_rebuild_full_params(self, param: FlatParameter):
                     )
             eod.param_order.append(param_index)
 
+    def register_comm_hook(self, state: object, hook: callable):
```
should we enforce to register hook on the root only?

aovladi replied:
Thanks for pointing that out! Added into the updated version.

zhaojuanmao reviewed (commented):
nice work! left some comments and questions

aovladi commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,172 @@
+        # Check that a Runtime Error is raised for other strategies
+        if sharding_strategy != ShardingStrategy.NO_SHARD:
+
+            with self.assertRaises(RuntimeError) as captured:
```
Thanks for pointing this out. Changed in the new commit.

aovladi commented on torch/distributed/algorithms/fsdp_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,49 @@
+    def __init__(
+        self,
+        process_group,
+        world_size
```
Thanks for corrections! Refactored this accordingly.

aovladi commented:
@rohan-varma  @zhaojuanmao @mrshenli Thanks for providing a feedback!
Since the first commit, I've addressed all comments and changed the logic:
as @mrshenli suggested, I've set `all_reduce` hook as a default one for a `NO_SHARD` strategy, for other strategies default hook is set to `None`. As a result, gradients are pre-divided if `comm_hook` is a default one(`None` or `all_reduce`). This change also required changes in unit testing. I've split tests in 2. One of them checks that registering of communication hooks raises proper errors and expected behavior. The second test, checks if `all_reduce` hook reduces gradients as expected.

Commit:
Fixes issues after rebase

Commit:
Fixed issues after merge

Commit:
Removed unnecessary logger

aovladi commented:
@pytorchbot rebase

Commit:
Fixed formatting for a docstring

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -927,6 +927,11 @@ def _run_param_init_fn():
+        # setting communication hook to a default
+        self.communication_hook = self._get_default_comm_hook()
+        self.communication_hook_state = self._get_default_comm_hook_state()
+        self._comm_hook_was_called = False
```
nit: comm_hook_was_registered might be a better name?

aovladi replied:
changed to "_hook_registered"

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -4018,6 +4024,69 @@ def rekey_optim_state_dict(
+            for submodule in self.fsdp_modules(self):
+                submodule._comm_hook_was_called = True
+                # registering hook only if it hasn't been already registered
+                if submodule.communication_hook == self._get_default_comm_hook():
```
double registration on submodule should still raise an error? maybe the code should be:

```
for submodule in self.fsdp_modules(self):
    assert not submodule._hook_registered
    # set the hooks
    submodule._hook_registered = True
    
```

aovladi replied:
Right, if a hook was registered on a submodule earlier, it won't catch it. Addressed that

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -4018,6 +4024,69 @@ def rekey_optim_state_dict(
+        assert self.check_is_root(), "register_comm_hook can only be called on a root instance."
+        assert not self._comm_hook_was_called, "register_comm_hook can be only called once"
+        if self.sharding_strategy != ShardingStrategy.NO_SHARD:
+            raise NotImplementedError(
```
would be nice to file an issue to implement it for the rest of the sharding strategies.

aovladi replied:
I will prepare an issue

Commit:
Enforced comm hook can be registered only once on submodules.

Commit:
Enforced comm hook can be registered only once on submodules.

Commit:
Added extra testing for comm_hooks on submodules

rohan-varma commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,239 @@
+                    if not submodule.check_is_root()
+                ]
+                # Check that assertion is raised for registering a comm hook on a non-root
+                with self.assertRaises(AssertionError):
```
nit: use `assertRaisesRegex` with the error message substr, otherwise the test can pass but with the wrong error raised, in theory

rohan-varma commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,239 @@
+                submodules[1]._hook_registered = True
+
+                # Check that an error is raised when some of submodules have a non-default hook assigned
+                with self.assertRaises(AssertionError):
```
ditto

rohan-varma commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,239 @@
+                dummy_hook
+            )
+            # Check that we can't register comm hook twice
+            with self.assertRaises(AssertionError):
```
ditto

rohan-varma commented on torch/distributed/algorithms/_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,59 @@
+        grad (torch.Tensor): A gradient for the local batch that needs to be communicated across ranks.
+    """
+    # `all_reduce` is a default hook and grad is already pre-divided
+    dist.all_reduce(grad, group=state.process_group)
```
looks like this hook expects a pre-divided gradient - would be good to document this clearly, that we expect predivision, but perform postdivision (also tbh that might be a little confusing to users but it's fine as this hook simulates existing FSDP behavior, we can add an allreduce hook that does not postdivide later).

Consider the above, might also be good to rename this hook to `allreduce_with_postdivide_hook`?

aovladi replied:
I've reconsidered the current behavior of a default `all_reduce` hook and in the updated version it takes care of pre- and post-division. This way there is no necessity in different `all_reduce` hooks and confusion is eliminated. Note: as was discussed earlier, `FSDP. _post_backward_hook` does not use sharding strategy explicitly for pre-division of gradients.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3397,6 +3402,7 @@ def _check_rebuild_full_params(self, param: FlatParameter):
                     )
             eod.param_order.append(param_index)
 
+
```
nit: extra line

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2897,7 +2901,8 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                     param.grad.data = param.grad.data.to(self.mixed_precision.reduce_dtype)
 
-                if self.gradient_predivide_factor > 1:
+                if self.gradient_predivide_factor > 1 and \
```
I see, so default comm hook is none when sharded, and the default hook when unsharded, so the old behavior is exactly retained

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -4018,6 +4024,69 @@ def rekey_optim_state_dict(
+                            This function takes in a python tensor, which represents a gradient
+                            of a loss function with respect to variables of a model
+                            for the local batch that needs to be communicated across ranks.
+                            It then performs all neccessary processing and returns nothing.
```
nit: returns `None`

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -4018,6 +4024,69 @@ def rekey_optim_state_dict(
+            hook (callable): Callable with the following signature:
+                            ``hook: Callable[torch.Tensor] -> None``:
+                            This function takes in a python tensor, which represents a gradient
+                            of a loss function with respect to variables of a model
```
maybe change this to: "...which represents the full, flattened, unsharded gradient with respect to all variables corresponding to the model this FSDP unit is wrapping (that are not wrapped by other FSDP sub-units")

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -4018,6 +4024,69 @@ def rekey_optim_state_dict(
+                assert not submodule._hook_registered, "communication hook can be only registered once"
+                submodule._hook_registered = True
+                # registering hook only if it hasn't been already registered
+                if submodule.communication_hook == self._get_default_comm_hook():
```
IIUC, this can be changed from an `if` to an `assert` right? 

Since we've checked above that `register_comm_hook` has not been called, so the only thing `submodule.comm_hook` should be able to return is `self.default_hook` (allreduce for no_shard and None for others).

so we should be able to change this line to `assert submodule.comm_hook == self._get_default_hook()`?

aovladi replied:
That is correct, refactored

rohan-varma reviewed (commented):
Nice job addressing all of the comments! Getting very close, just some minor comments / suggestions, will stamp after those are resolved

awgu commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,239 @@
+            loss.backward()
+
+            # For each worker, the gradient on the weight should be worker_rank.
+            grad = [param.grad for param in net_default_hook.parameters()]
```
nit:
```suggestion
            grad = net_default_hook.params[0].grad
```
The variable `grad` being `list` may be misleading since it is always a singleton list. (The code already assumes that there is a single `FlatParameter` in `net_default_hook`, which is a valid assumption because there is only a single top-level FSDP instance.)

aovladi replied:
Got it. refactored

awgu commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,239 @@
+        inpt = torch.tensor([self.rank]).float().cuda(self.rank)
+
+        net_default_hook = FSDP(
+            copy.deepcopy(m),
```
nit: No need to deep copy I think.

awgu commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,239 @@
+        """
+        Tests FSDP's default communication hook's behaviour and correctness.
+        Arguments:
+            has_wrapping (bool): Configures wrapping of a module.
```
```suggestion
```

awgu commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,239 @@
+            # For each worker, the gradient on the weight should be worker_rank.
+            grad = [param.grad for param in net_default_hook.parameters()]
+
+            avg = copy.deepcopy(grad[0])
```
nit: No need to deep copy I think.

awgu commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,239 @@
+            )
+            # Verify default hook produces expected gradients
+            self.assertEqual(
+                avg[0].item(),
```
noob question: Why are we taking `avg[0].item()` here? I may be misreading, but it looks like `grad = [flat_param.grad]` and `avg = grad[0] = flat_param.grad`. Does this mean that `avg[0]` is the 0th element of `avg`, returned as a `Tensor`? Is there any need to call `avg[0].item()`, or can we just call `avg.item()`?

aovladi replied:
`avg.item()` won't work, because `avg` has multiple elements. Test works in such a way, that gradient values are the same, i.e. in this case for 1X2 network `grad` would be `tensor([0.5000, 0.5000])`. Thus, we only need to compare one of these values with the expected one.

awgu commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,239 @@
+        dummy_state = DummyState(process_group=None)
+
+        # FSDP currently suports communication hooks for a NO_SHARD strategy
+        # Check that a `NotImplementedError`` is raised for other strategies
```
nit: Inconsistent backtick

awgu commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,239 @@
+
+            # Check that default hook is set to `all_reduce`
+            for entry in FSDP.fsdp_modules(fsdp_model_with_hook):
+                self.assertEqual(entry.communication_hook.__qualname__, allreduce_hook.allreduce_hook.__qualname__)
```
Curious: Is it possible to check the function reference itself?
```suggestion
                self.assertEqual(entry.communication_hook, allreduce_hook.allreduce_hook)
```

awgu commented on torch/distributed/algorithms/_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,59 @@
+
+class AllReduceState(object):
+    r"""
+    Stores parameters, needed to perform ``all_reduce`` algorithm
```
nit:
```suggestion
    Stores state needed to perform the ``all_reduce`` algorithm
```
(just a suggestion)

awgu commented on torch/distributed/algorithms/_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,59 @@
+
+    Args:
+        process_group (ProcessGroup): The process group to be used for all-reduce.
+        world_size (int): The number of nodes in a process group.
```
There may be multiple workers per node (e.g. 8 GPUs per 1 node in a multi-node cluster).
```suggestion
        world_size (int): The number of workers in a process group.
```

awgu commented on torch/distributed/algorithms/_comm_hooks/allreduce_hook.py:
```diff
@@ -0,0 +1,59 @@
+        )
+        self.gradient_postdivide_factor = self.world_size / self.gradient_predivide_factor
+
+    # setting two factors 'self.gradient_predivide_factor'
```
nit: ' -> `

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -4018,6 +4024,69 @@ def rekey_optim_state_dict(
+        across multiple workers.
+        This hook can be used to implement several algorithms like GossipGrad
+        and gradient compression which involve different communication strategies for
+        parameter syncs while running Fully Sharded Data Parallel training.
```
nit: 
```suggestion
        parameter syncs while training with :class:`FullyShardeDataParallel`.
```

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2897,7 +2901,8 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
 
-                if self.gradient_predivide_factor > 1:
+                if self.gradient_predivide_factor > 1 and \
+                        self.communication_hook == self._get_default_comm_hook():
```
This logic is pretty tricky. The specification this implies is that we only pre-divide the gradient when using the default communication hook. However, how can a user pre-divide the gradient otherwise? My understanding is that it should be done inside the hook.

Would it be better to pre-divide the gradient in the new all-reduce hook for `NO_SHARD` to be consistent with all registered hooks?

Then, the logic looks like:
- If you are using `NO_SHARD`, then the communication hook is *always* responsible for the pre-division, even if it is the default all-reduce hook.
- If you are using another strategy, then FSDP will pre-divide for you since there is no way for you to pre-divide yourself.

(In short, I do not think we should treat the `NO_SHARD` default hook as a special case and only pre-divide for it but not other hooks.)

aovladi replied:
@awgu I see what you mean. @rohan-varma and @zhaojuanmao mentioned earlier (in previous version of this PR), that we should base pre-division logic on the sharding strategy. For this iteration, I based my logic on whether a default hook is picked or not to retain the default behavior. That being said, I use `communication_hook` as a proxy to determine a sharding strategy:  if `communication_hook` is `None` ( which is true for sharded strategies). Then, I believe when communication hooks are implemented for all sharding strategies, pre-division can be done solely in hooks (i.e. get rid of pre-division here).

awgu replied:
I might be missing something. Why do we not do something like:
```
if self.sharding_strategy != ShardingStrategy.NO_SHARD:
    param.grad.div_(self.gradient_predivide_factor)
```
and then add the `param.grad.div_(self.gradient_predivide_factor)` as well in the default `NO_SHARD` communication hook before the all-reduce?

Does this not preserve the existing behavior as well?

aovladi replied:
Referencing an earlier discussion: https://github.com/pytorch/pytorch/pull/79833#discussion_r900703089

aovladi replied:
I've refactored `all_reduce` to have pre-division. As for general logic in pre-division, I believe even though doing a pre-division based on a sharding strategy  may preserve the behavior,  the first question would be: Why specifically `NO_SHARD` is excluded - a context is needed (i.e. for `NO_SHARD` pre-division is implemented inside of its default hook). If we do pre-division based on a value of `self.communication_hook`, it at least gives a hint, that pre-division is implemented inside of a `self.communication_hook` when it is not `None`. Since at that point for `SHARD_GRAD_OP` and `FULL_SHARD` default hook is `None`, it does give an idea why pre-division is still there in place.

awgu replied:
> Why specifically `NO_SHARD` is excluded - a context is needed (i.e. for `NO_SHARD` pre-division is implemented inside of its default hook). 

I am not sure I am following this part.

Specifically `NO_SHARD` is excluded because now the communication is refactored into a hook. Logically, everything related to the the gradient communication is _encapsulated_ into the hook. If we implement communication hooks for the other strategies, I expect those hooks will run the pre- and post-division as well.

It does not make sense to me why we would only attribute post-division to be part of FSDP's semantics and not the pre-division. It should be either both pre- and post-division or neither (especially when the two factors are expected to be reciprocals). Right now, I prefer having both in the hook because there may exist communication algorithms for which pre-division does not apply (e.g. the communication algorithm's reduction does not scale with the world size).

To elaborate, do we want to enforce that all hooks have a state and that that state has a `gradient_predivide_factor` attribute? That seems restrictive. Hence, if we do the pre-division only for the default hook, then we are hard coding the fact that the default hook _does_ have a `gradient_predivide_factor` attribute, which is brittle.

Your previous implementation looked like (before the recent commits):
```
if self.gradient_predivide_factor > 1 and \
        self.communication_hook == self._get_default_comm_hook():
    # Average grad by world_size for consistency with PyTorch DDP.
    param.grad.div_(self.gradient_predivide_factor)
```
This actually uses the FSDP instance's `self.gradient_predivide_factor` and not the one from the `NO_SHARD` default hook. If this was corrected, you would see how you have to assume that `hasattr(self.communication_hook, "gradient_predivide_factor")`.

> If we do pre-division based on a value of `self.communication_hook`, it at least gives a hint, that pre-division is implemented inside of a `self.communication_hook` when it is not `None`. Since at that point for `SHARD_GRAD_OP` and `FULL_SHARD` default hook is `None`, it does give an idea why pre-division is still there in place.

To clarify, is this part explaining why now the check for FSDP's pre-division is `if self.communication_hook is None`?
If so, I think that checking `self.communication_hook` against `None` is fine. Would I be correct to say that it is currently equivalent to checking if the sharding strategy is not `NO_SHARD`?

aovladi replied:
@awgu I agree with you on all of your concerns and previous logic was revised. I do agree communication hooks should take care of both pre- and post-division (if any is required). 

> Specifically NO_SHARD is excluded because now the communication is refactored into a hook. 

I was trying to say, that I know this context, but not everyone does.

>Logically, everything related to the the gradient communication is encapsulated into the hook. If we implement communication hooks for the other strategies, I expect those hooks will run the pre- and post-division as well.

I apologize, that my previous version of PR caused a confusion, but I did have the same understanding in mind. I do understand where confusion comes from and I will avoid it in future. 

> Hence, if we do the pre-division only for the default hook, then we are hard coding the fact that the default hook does have a gradient_predivide_factor attribute, which is brittle.

Previous implementation used FSDP instance's `self.gradient_predivide_factor` to pre-divide gradients and then the default hook post-divided them with a recomputed `gradient_postdivide_factor` (stored in its state). This split was unnecessary, I agree. 

Current pre-division for `SHARD_GRAD_OP` and `FULL_SHARD` strategies  is left in place to preserve a correct behavior, because these strategies don't have a default hook implemented for them. My understanding was, that we can fully transfer this pre-division into default hooks once all of them are implemented and we can get rid of this pre-division here (which I failed to explain, I apologize). Getting rid of pre-division at that point, would also introduce changes to a default `all_reduce` hook (i.e. adding pre-division).

> To clarify, is this part explaining why now the check for FSDP's pre-division is if self.communication_hook is None?

I tried my best to explain that.

> Would I be correct to say that it is currently equivalent to checking if the sharding strategy is not NO_SHARD?

Currently, yes.

awgu replied:
Thanks for the clarifications! One last question:

> Getting rid of pre-division at that point, would also introduce changes to a default all_reduce hook (i.e. adding pre-division).

I thought the default all-reduce hook now (in your PR) has the pre-division built in, so we do not need to add pre-division to it? It seems like we would not need to modify the default all-reduce hook when we add hooks for the other strategies.

aovladi replied:
I was trying to provide explanations for the previous version. Of course, since in current version pre-division is already in `all_reduce` we won't need to add an extra pre-division later, i.e. less things to keep in mind.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -4018,6 +4024,69 @@ def rekey_optim_state_dict(
+            # register same hook for root and all submodules
+            for submodule in self.fsdp_modules(self):
+                assert not submodule._hook_registered, "communication hook can be only registered once"
+                submodule._hook_registered = True
```
Curious: is there a way to track this information without an explicit `bool` (i.e. based on the other registered state)?

aovladi replied:
well, as a straightforward suggestion: when a hook is registered, `comm_hook` changes its state from `default` to a `new` one. Technically, we can judge based on current state, i.e. if current is anything, but `default`, we can't register. I do need to think about it more

awgu replied:
I see. It seems that there is a design choice here that is up to you. Either we can keep the `_hook_registered: bool`, or we can (1) check against default and (2) ensure that `register_comm_hook()` is idempotent for the default hook.

awgu replied:
Nevermind. I see that @rohan-varma suggested we error if we double-register a hook, even if it is the default one (violating idempotence).

awgu reviewed (commented):
Nice work! I made an initial pass. Feel free to request another review later.

Commit:
Transfered pre-division to all_reduce hook. Addressed comments.

Commit:
Added small correction to test in regards to functions' assert

Commit:
Added missing regex to assertion check

Commit:
Brought back boolean flag

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2897,7 +2901,7 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
 
-                if self.gradient_predivide_factor > 1:
+                if self.gradient_predivide_factor > 1 and self.communication_hook is None:
                     # Average grad by world_size for consistency with PyTorch DDP.
```
(This is prior to the PR) The comment does not seem to match what we do here? `gradient_predivide_factor` might or might not be the `world_size`?

rohan-varma replied:
Agreed, the process of pre + post division together is what leads overall to the averaging by world size. Would be good to send a separate PR updating the docs to reflect this.

aovladi replied:
Sounds good, I will take a look and prepare a separate PR

mrshenli reviewed (commented):
LGTM! Thanks! Will leave it to @rohan-varma and @awgu to stamp when their comments are addressed.

Commit:
Fixed typos in docstrings. Adjusted allreduce default hook's predivide.

rohan-varma reviewed (approved):
LGTM, looking forward to see the benchmark results (accuracy + QPS) as well! Thanks for working on this!

awgu commented on test/distributed/fsdp/test_fsdp_comm_hooks.py:
```diff
@@ -0,0 +1,253 @@
+                self.assertEqual(
+                    entry.communication_hook_state,
+                    dummy_state
+                )
```
Something to consider: This one test `test_default_communication_hook_initialization` is testing quite a bit even though they may all be considered under the "hook registration interface": e.g. non-supported sharding strategies should error, submodule versus root module hook parity, and not registering a hook twice.

It could be worthwhile to break this test up (though it is up to you). If one assertion fails, I think pytest will terminate the test, which is why it is good to discretize unit tests. Of course, the granularity for discretization is subjective.

aovladi replied:
Valid point. I've incorporated granularity into current tests.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -4018,6 +4022,71 @@ def rekey_optim_state_dict(
+                            and shared by all the gradient tensors on the worker.
+            hook (callable): Callable with the following signature:
+                            ``hook: Callable[torch.Tensor] -> None``:
+                            This function takes in a python tensor, which represents
```
nit: 
```suggestion
                            This function takes in a Python tensor, which represents
```

awgu reviewed (commented):
Nice job with this. I am leaving a final round of comments. Feel free to address them as you please.

aovladi commented:
Benchmark results were linked to internal diff

aovladi commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -4018,6 +4022,71 @@ def rekey_optim_state_dict(
+                            It then performs all necessary processing and returns ``None``.
+
+        """
+        assert self.check_is_root(), "register_comm_hook can only be called on a root instance."
```
Got it, thanks for this feedback, I've incorporated changes here, and will consider this factor in future.

Commit:
Addressed comments. Made tests more granular.

awgu reviewed (approved):
Awesome work! This looks really good now.

aovladi commented:
@pytorchbot rebase

aovladi commented:
@pytorchbot merge -g

aovladi commented:
@pytorchbot rebase

aovladi commented:
@pytorchbot merge -g

aovladi commented:
@pytorchbot rebase -s

aovladi commented:
@pytorchbot rebase

aovladi commented:
@pytorchbot merge -g

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2974,10 +2978,10 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
-                            param.grad.div_(self.gradient_postdivide_factor)
+                        # if a communication hook was not registered,
+                        # then a default hook (`all_reduce`) will be used
+                        self.communication_hook(self.communication_hook_state, param.grad)
```
nit: in your next PR, would you please assert 'self.communication_hook' and 'self.communication_hook_state' is not None as well before calling them?

aovladi replied:
sure, thank you for pointing that out

zhaojuanmao commented:
since the NO_SHARD implementation is replaced with communication_hook() now, wondering how is the benchmarks result compared to trunk?

aovladi commented:
@zhaojuanmao Benchmark results were shared in the internal diff [on Phabricator](https://www.internalfb.com/diff/D37397210).
